WcaOnRails/app/models/user.rb

```diff
@@ -692,8 +702,14 @@ def can_add_and_remove_events?(competition)
     can_admin_competitions? || (can_manage_competition?(competition) && !competition.confirmed?)
   end
 
+  def can_upload_competition_results?(competition)
+    appropriate_role = can_admin_results? || competition.delegates.include?(self) || competition.trainee_delegates.include?(self)
+    appropriate_time = competition.in_progress? || competition.is_probably_over?
+    appropriate_role && appropriate_time && !competition.results_posted?
```

Comment:
This is the reason of your failing tests!
You allow the upload of competition results for unannounced competitions; the method below has been patched recently but you copy pasted code before the patch.

Whenever you copy paste some codes, it almost always mean you can "DRY" (Don't Repeat Yourself) it.

Here I would suggest to do something like that:
```ruby
def can_submit_competition_results?(competition, upload_only: false)
  allowed_delegate = if upload_only ?
                       competition.delegates.include?(self) || competition.trainee_delegates.include?(self)
                     else
                       competition.delegates.include?(self) && !competition.trainee_delegates.include?(self))
                     end
  appropriate_role = can_admin_results? || allowed_delegate
  appropriate_time = competition.in_progress? || competition.is_probably_over?
  competition.announced? && appropriate_role && appropriate_time && !competition.results_posted?
end

def can_upload_competition_results?(competition)
  can_submit_competition_results?(competition, upload_only: true)
end
```

This way you avoid any code duplication and make it clear to the reader that the only difference between a submission and an upload is the delegate status.

The exact same comment applies to the `can_post_delegate_report?`: could you please refactor the code the same way?

Reply:
Wow this looks cool, I'll do this.